File: python/03_signature_scoring.py
"""
Step 3: Signature scoring analysis using ssGSEA (Python implementation)
Correlate METRN expression with 7 pathway signature scores in GSE124272
"""

import numpy as np
import pandas as pd
from scipy import stats
import os
import warnings
warnings.filterwarnings('ignore')

# Try gsva package; if not working, use manual ssGSEA
out_dir = "D:/MedResearch/METRN_IVDD"

# ---- Load gene-level expression matrix from R output ----
# We have the FDR CSV files from Step 1 which have probe-level data
# But we need gene-level. Let's reconstruct from the series matrix.

# Since R is unstable, use GEOparse in Python
import GEOparse

gse1 = GEOparse.get_GEO(geo="GSE124272", destdir=os.path.join(out_dir, "geo_cache"), silent=True)

# Extract expression matrix
expr1 = gse1.pivot_samples('VALUE').T  # samples as rows, probes as columns
# Actually GEOparse gives probes as index, samples as columns
expr1_raw = gse1.pivot_samples('VALUE')
print(f"Expression matrix: {expr1_raw.shape}")

# Map probes to gene symbols
probe_to_gene = {}
for gsm_name, gsm in gse1.gsms.items():
    pass  # Not needed, use table_data

# Use the platform annotation
gpl = gse1.gpls[gse1.metadata['platform_id'][0]]
print(f"Platform: {gpl.metadata['platform_id']}")
print(f"GPL table columns: {list(gpl.table.columns)}")

# Create probe-to-gene mapping
if 'GENE_SYMBOL' in gpl.table.columns:
    probe_gene_map = dict(zip(gpl.table['ID'], gpl.table['GENE_SYMBOL']))
elif 'Gene Symbol' in gpl.table.columns:
    probe_gene_map = dict(zip(gpl.table['ID'], gpl.table['Gene Symbol']))
else:
    # Search for gene symbol column
    for col in gpl.table.columns:
        if 'gene' in col.lower() and 'symbol' in col.lower():
            probe_gene_map = dict(zip(gpl.table['ID'], gpl.table[col]))
            print(f"Using column: {col}")
            break

# Filter out empty gene symbols
probe_gene_map = {k: v for k, v in probe_gene_map.items() if v and v.strip() != ''}

# Collapse probes to genes (median of multiple probes)
print(f"Mapped probes: {len(probe_gene_map)}")

# Create gene-level expression
gene_expr_dict = {}
for probe, gene in probe_gene_map.items():
    if probe in expr1_raw.index:
        vals = expr1_raw.loc[probe].values
        if gene in gene_expr_dict:
            # Take median of multiple probes
            existing = gene_expr_dict[gene]
            gene_expr_dict[gene] = np.median([existing, vals], axis=0)
        else:
            gene_expr_dict[gene] = vals

# Build gene expression DataFrame
sample_names = expr1_raw.columns
gene_expr = pd.DataFrame(gene_expr_dict, index=sample_names).T
print(f"Gene-level matrix: {gene_expr.shape}")

# ---- Define gene sets ----
gene_sets = {
    'Mechanotransduction': ['PIEZO1', 'PIEZO2', 'TRPV4', 'YAP1', 'WWTR1', 
        'CTGF', 'CYR61', 'ITGA1', 'ITGA2', 'ITGA3', 'ITGA5', 'ITGA7',
        'ITGB1', 'ITGB3', 'ITGB5', 'ITGB6',
        'PTK2', 'RHOA', 'ROCK1', 'ROCK2', 'LIMK1', 'CFL1', 
        'VCL', 'TLN1', 'FOSB', 'CREB3L2', 'TMEM102'],
    
    'ECM_Homeostasis': ['COL2A1', 'ACAN', 'SOX9', 'COL1A1', 'FN1', 'COMP', 
        'MMP1', 'MMP3', 'MMP13', 'ADAMTS4', 'ADAMTS5', 'COL10A1', 'COL3A1'],
    
    'Autophagy': ['ATG5', 'ATG7', 'ATG12', 'ULK1', 'BECN1', 
        'MAP1LC3B', 'SQSTM1', 'ATG3', 'ATG4B', 'ATG16L1'],
    
    'Senescence': ['CDKN1A', 'CDKN2A', 'TP53', 'IL6', 'CXCL8', 
        'MMP1', 'MMP3', 'MMP13', 'SERPINE1', 'RB1', 'CCND1'],
    
    'ER_Stress': ['CALR', 'HSPA5', 'ATF4', 'DDIT3', 'XBP1', 
        'ERN1', 'CREB3L2', 'HSP90B1', 'PDIA3', 'EIF2AK3'],
    
    'Inflammation': ['IL1B', 'IL6', 'TNF', 'NFKB1', 'CXCL8', 
        'CCL2', 'TLR2', 'TLR4', 'IL10', 'TNFAIP3'],
    
    'Ferroptosis': ['GPX4', 'SLC7A11', 'ACSL4', 'FTH1', 'TFRC', 
        'NCOA4', 'ALOX15', 'POR', 'FDFT1', 'HSPB1']
}

# Filter gene sets for available genes
available_genes = set(gene_expr.index)
filtered_sets = {}
for name, genes in gene_sets.items():
    found = [g for g in genes if g in available_genes]
    missing = [g for g in genes if g not in available_genes]
    print(f"{name}: found {len(found)}/{len(genes)} genes, missing: {missing}")
    if len(found) >= 2:
        filtered_sets[name] = found

print(f"\nUsable gene sets: {list(filtered_sets.keys())}")

# ---- Manual ssGSEA implementation ----
def ssgsea_score(expr_matrix, gene_set):
    """
    Simplified ssGSEA: rank-based enrichment score
    For each sample, rank genes by expression, then compute 
    running sum of gene set membership.
    """
    scores = {}
    for sample in expr_matrix.columns:
        # Rank genes by expression (descending)
        sample_vals = expr_matrix[sample].sort_values(ascending=False)
        ranks = pd.Series(range(1, len(sample_vals) + 1), index=sample_vals.index)
        
        # Compute enrichment score
        N = len(ranks)
        Nh = len([g for g in gene_set if g in ranks.index])
        if Nh == 0:
            scores[sample] = 0
            continue
        
        # Running sum
        hit_indices = ranks.index.isin(gene_set)
        running_sum = 0
        max_sum = 0
        
        for i, gene in enumerate(ranks.index):
            if gene in gene_set:
                running_sum += ranks[gene] / sum(ranks[gene] for g in gene_set if g in ranks.index)
            else:
                running_sum -= 1.0 / (N - Nh)
            if abs(running_sum) > abs(max_sum):
                max_sum = running_sum
        
        scores[sample] = max_sum
    
    return pd.Series(scores)

# Alternative: simple mean-based scoring (more robust for small samples)
def mean_signature_score(expr_matrix, gene_set):
    """Mean z-score of gene set members"""
    available = [g for g in gene_set if g in expr_matrix.index]
    if len(available) < 2:
        return pd.Series(0, index=expr_matrix.columns)
    
    subset = expr_matrix.loc[available]
    # Z-score each gene across samples, then average
    z_scores = (subset - subset.mean(axis=1).values.reshape(-1, 1)) / subset.std(axis=1).values.reshape(-1, 1)
    # Replace NaN (constant genes) with 0
    z_scores = z_scores.fillna(0)
    return z_scores.mean(axis=0)

# ---- Compute signature scores ----
print("\n=== Computing signature scores ===")
signature_scores = pd.DataFrame(index=gene_expr.columns)
for name, genes in filtered_sets.items():
    # Use mean z-score method (more robust)
    scores = mean_signature_score(gene_expr, genes)
    signature_scores[name] = scores
    print(f"{name}: computed")

print(f"Signature scores shape: {signature_scores.shape}")

# ---- Correlate METRN with signature scores ----
print("\n=== METRN-Signature correlations ===")
metrn_vals = gene_expr.loc['METRN']

cor_results = []
for sig_name in signature_scores.columns:
    sig_vals = signature_scores[sig_name]
    r, p = stats.spearmanr(metrn_vals.values, sig_vals.values)
    cor_results.append({
        'Dataset': 'GSE124272',
        'Signature': sig_name,
        'Spearman_r': r,
        'P_value': p,
        'n_genes': len(filtered_sets[sig_name])
    })
    print(f"{sig_name}: r={r:.3f}, P={p:.4f}")

cor_df = pd.DataFrame(cor_results)

# FDR correction
from scipy.stats import false_discovery_control
cor_df['FDR'] = false_discovery_control(cor_df['P_value'].values, method='bh') if len(cor_df) > 0 else [1.0]*len(cor_df)

print("\nWith FDR correction:")
print(cor_df.to_string())

# ---- METRN-PIEZO1 correlation ----
print("\n=== METRN-PIEZO1 correlation ===")
piezo1_vals = gene_expr.loc['PIEZO1']
r_all, p_all = stats.spearmanr(metrn_vals.values, piezo1_vals.values)
r_pearson, p_pearson = stats.pearsonr(metrn_vals.values, piezo1_vals.values)
print(f"All samples: Spearman r={r_all:.3f}, P={p_all:.4f}")
print(f"All samples: Pearson r={r_pearson:.3f}, P={p_pearson:.4f}")

# ---- Top co-expressed genes ----
print("\n=== Top METRN co-expressed genes ===")
gene_cors = []
for gene in gene_expr.index:
    if gene == 'METRN':
        continue
    vals = gene_expr.loc[gene].values
    if np.std(vals) == 0:
        continue
    r, p = stats.spearmanr(metrn_vals.values, vals)
    gene_cors.append({'Gene': gene, 'Spearman_r': r, 'P_value': p})

gene_cor_df = pd.DataFrame(gene_cors).sort_values('Spearman_r', key=abs, ascending=False)

# FDR for all gene correlations
if len(gene_cor_df) > 0:
    gene_cor_df['FDR'] = false_discovery_control(gene_cor_df['P_value'].values, method='bh')

# Top 20
top20 = gene_cor_df.head(20)
print("Top 20 co-expressed genes:")
print(top20.to_string())

# Key genes
key_genes = ['PIEZO1', 'CALR', 'LIMK1', 'FOSB', 'CREB3L2', 'TMEM102', 'MESP1', 'ATG12', 'ULK1', 'COL2A1', 'MMP13', 'SOX9']
print("\nKey gene correlations:")
for g in key_genes:
    row = gene_cor_df[gene_cor_df['Gene'] == g]
    if len(row) > 0:
        print(f"  {g}: r={row['Spearman_r'].values[0]:.3f}, P={row['P_value'].values[0]:.4f}, FDR={row['FDR'].values[0]:.4f}")
    else:
        print(f"  {g}: not found")

# ---- Save all results ----
print("\n=== Saving results ===")
cor_df.to_csv(os.path.join(out_dir, 'signature_correlation_results.csv'), index=False)
signature_scores.to_csv(os.path.join(out_dir, 'GSVA_scores_GSE124272.csv'))
gene_cor_df.to_csv(os.path.join(out_dir, 'METRN_coexpression_all_genes_GSE124272.csv'))
top20.to_csv(os.path.join(out_dir, 'METRN_top20_coexpressed_GSE124272.csv'), index=False)

# ---- Generate plots ----
print("\n=== Generating plots ===")
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

# Bar plot of METRN-signature correlations
fig, ax = plt.subplots(figsize=(8, 5))
sig_names = cor_df['Signature'].values
r_vals = cor_df['Spearman_r'].values
fdr_vals = cor_df['FDR'].values

colors = ['#F44336' if f < 0.05 else '#9E9E9E' for f in fdr_vals]
bars = ax.bar(range(len(sig_names)), r_vals, color=colors)
ax.set_xticks(range(len(sig_names)))
ax.set_xticklabels(sig_names, rotation=45, ha='right')
ax.axhline(y=0, color='gray', linestyle='--')
ax.set_ylabel('Spearman r')
ax.set_title('METRN-Signature Score Correlations (GSE124272)')
# Add legend
from matplotlib.patches import Patch
legend_elements = [Patch(facecolor='#F44336', label='FDR < 0.05'),
                   Patch(facecolor='#9E9E9E', label='FDR >= 0.05')]
ax.legend(handles=legend_elements, loc='upper right')

plt.tight_layout()
plt.savefig(os.path.join(out_dir, 'METRN_signature_correlation_GSE124272.png'), dpi=150)
plt.savefig(os.path.join(out_dir, 'METRN_signature_correlation_GSE124272.pdf'))
print("Bar plot saved")

# METRN-PIEZO1 scatter
fig, ax = plt.subplots(figsize=(7, 5))
# Get labels
pdata = gse1.phenotype_data
labels = []
for gsm in metrn_vals.index:
    title = pdata.loc[gsm, 'title'] if gsm in pdata.index else ''
    if 'patient' in title.lower() or 'IDD' in title:
        labels.append('IDD')
    elif 'volunteer' in title.lower() or 'healthy' in title.lower() or 'control' in title.lower():
        labels.append('Control')
    else:
        labels.append('Unknown')

idd_mask = [l == 'IDD' for l in labels]
ctrl_mask = [l == 'Control' for l in labels]

ax.scatter(metrn_vals.values[idd_mask], piezo1_vals.values[idd_mask], c='#F44336', s=60, label='IDD', zorder=3)
ax.scatter(metrn_vals.values[ctrl_mask], piezo1_vals.values[ctrl_mask], c='#4CAF50', s=60, label='Control', zorder=3)

# Add regression line
from numpy.polynomial.polynomial import polyfit
b, m = polyfit(metrn_vals.values, piezo1_vals.values, 1)
x_line = np.linspace(metrn_vals.values.min(), metrn_vals.values.max(), 100)
ax.plot(x_line, b + m * x_line, 'k-', linewidth=1)

ax.set_xlabel('METRN Expression')
ax.set_ylabel('PIEZO1 Expression')
ax.set_title(f'METRN vs PIEZO1 (GSE124272)\nSpearman r={r_all:.3f}, P={p_all:.4f}')
ax.legend()

plt.tight_layout()
plt.savefig(os.path.join(out_dir, 'METRN_PIEZO1_scatter_GSE124272.png'), dpi=150)
plt.savefig(os.path.join(out_dir, 'METRN_PIEZO1_scatter_GSE124272.pdf'))
print("Scatter plot saved")

print("\n=== Step 3 Complete ===")
print(f"All output saved to: {out_dir}")

File: R/03_signature_scoring.R
#!/usr/bin/env Rscript
# ============================================================
# Step 3: Signature scoring analysis using GSVA
# ============================================================

library(GEOquery)
library(limma)
library(GSVA)
library(GSEABase)
library(ggplot2)
library(dplyr)
library(reshape2)

out_dir <- "D:/MedResearch/METRN_IVDD"
geo_cache <- file.path(out_dir, "geo_cache")

# ---- Define gene sets ----
cat("=== Defining gene sets ===\n")

mechanotransduction_genes <- c("PIEZO1", "PIEZO2", "TRPV4", "YAP1", "WWTR1", 
  "CTGF", "CYR61", "ITGA1", "ITGA2", "ITGA3", "ITGA5", "ITGA7", "ITGA10", "ITGA11",
  "ITGB1", "ITGB3", "ITGB5", "ITGB6", "ITGB8",
  "PTK2", "RHOA", "ROCK1", "ROCK2", "LIMK1", "CFL1", 
  "VCL", "TLN1", "FOSB", "CREB3L2", "TMEM102")

ecm_genes <- c("COL2A1", "ACAN", "SOX9", "COL1A1", "FN1", "COMP", 
  "MMP1", "MMP3", "MMP13", "ADAMTS4", "ADAMTS5", "COL10A1", "COL3A1")

autophagy_genes <- c("ATG5", "ATG7", "ATG12", "ULK1", "BECN1", 
  "MAP1LC3B", "SQSTM1", "ATG3", "ATG4B", "ATG4D", "ATG10", "ATG16L1")

senescence_genes <- c("CDKN1A", "CDKN2A", "TP53", "IL6", "CXCL8", 
  "MMP1", "MMP3", "MMP13", "SERPINE1", "RB1", "CCND1")

er_stress_genes <- c("CALR", "HSPA5", "ATF4", "DDIT3", "XBP1", 
  "ERN1", "CREB3L2", "HSP90B1", "PDIA3", "EIF2AK3")

inflammation_genes <- c("IL1B", "IL6", "TNF", "NFKB1", "CXCL8", 
  "CCL2", "TLR2", "TLR4", "IL10", "TNFAIP3")

ferroptosis_genes <- c("GPX4", "SLC7A11", "ACSL4", "FTH1", "TFRC", 
  "NCOA4", "ALOX15", "POR", "FDFT1", "HSPB1")

# Combine into gene set list
gene_sets <- list(
  Mechanotransduction = mechanotransduction_genes,
  ECM_Homeostasis = ecm_genes,
  Autophagy = autophagy_genes,
  Senescence = senescence_genes,
  ER_Stress = er_stress_genes,
  Inflammation = inflammation_genes,
  Ferroptosis = ferroptosis_genes
)

cat("Gene set sizes:\n")
for (nm in names(gene_sets)) {
  cat(nm, ": ", length(gene_sets[nm]), " genes\n")
}

# ---- Load expression data ----
cat("\n=== Loading expression data ===\n")

# GSE124272 (IVDD peripheral blood) - main analysis dataset
gse1 <- getGEO(filename = file.path(geo_cache, "GSE124272_series_matrix.txt.gz"))
expr1 <- exprs(gse1)
pdata1 <- pData(gse1)
fd1 <- fData(gse1)

# GSE23130 (IVDD disc tissue)
gse3 <- getGEO(filename = file.path(geo_cache, "GSE23130_series_matrix.txt.gz"))
expr3 <- exprs(gse3)
pdata3 <- pData(gse3)
fd3 <- fData(gse3)

cat("GSE124272:", nrow(expr1), "probes,", ncol(expr1), "samples\n")
cat("GSE23130:", nrow(expr3), "probes,", ncol(expr3), "samples\n")

# ---- Map probes to gene symbols ----
cat("\n=== Mapping probes to genes ===\n")

map_to_genes <- function(expr_mat, fdata) {
  # Use GENE_SYMBOL column
  gene_col <- "GENE_SYMBOL"
  if (!gene_col %in% colnames(fdata)) {
    for (col in colnames(fdata)) {
      if (grepl("gene|symbol", col, ignore.case = TRUE)) {
        gene_col <- col
        break
      }
    }
  }
  cat("Using gene column:", gene_col, "\n")
  
  # Create gene-to-probe mapping
  fdata$gene_sym <- fdata[[gene_col]]
  # Remove empty gene symbols
  fdata_filtered <- fdata[fdata$gene_sym != "" & !is.na(fdata$gene_sym), ]
  
  # Collapse multiple probes per gene by median
  expr_filtered <- expr_mat[fdata_filtered$ID, ]
  
  # Group by gene symbol
  gene_expr <- t(sapply(unique(fdata_filtered$gene_sym), function(gene) {
    probes <- fdata_filtered$ID[fdata_filtered$gene_sym == gene]
    if (length(probes) == 1) {
      return(expr_filtered[probes, ])
    } else {
      return(apply(expr_filtered[probes, ], 2, median))
    }
  }))
  
  rownames(gene_expr) <- unique(fdata_filtered$gene_sym)
  return(gene_expr)
}

gene_expr1 <- map_to_genes(expr1, fd1)
cat("GSE124272 gene-level matrix:", nrow(gene_expr1), "genes,", ncol(gene_expr1), "samples\n")

gene_expr3 <- map_to_genes(expr3, fd3)
cat("GSE23130 gene-level matrix:", nrow(gene_expr3), "genes,", ncol(gene_expr3), "samples\n")

# ---- Find METRN in gene-level matrix ----
cat("\n=== Finding METRN ===\n")
metrn1 <- gene_expr1["METRN", ]
cat("GSE124272 METRN found:", !is.null(metrn1), "\n")
if (!is.null(metrn1)) cat("METRN values:", round(metrn1, 3), "\n")

metrn3 <- gene_expr3["METRN", ]
cat("GSE23130 METRN found:", !is.null(metrn3), "\n")

# ---- Filter gene sets for available genes ----
cat("\n=== Checking gene set coverage ===\n")

filter_geneset <- function(gene_set, available_genes) {
  found <- intersect(gene_set, available_genes)
  missing <- setdiff(gene_set, available_genes)
  cat("  Found:", length(found), "/", length(gene_set), "genes\n")
  if (length(missing) > 0) cat("  Missing:", missing, "\n")
  return(found)
}

# GSE124272
filtered_sets1 <- lapply(gene_sets, filter_geneset, rownames(gene_expr1))
# Remove empty sets
filtered_sets1 <- filtered_sets1[sapply(filtered_sets1, length) > 0]
cat("Usable gene sets for GSE124272:", names(filtered_sets1), "\n")

# GSE23130
filtered_sets3 <- lapply(gene_sets, filter_geneset, rownames(gene_expr3))
filtered_sets3 <- filtered_sets3[sapply(filtered_sets3, length) > 0]
cat("Usable gene sets for GSE23130:", names(filtered_sets3), "\n")

# ---- GSVA analysis ----
cat("\n=== Running GSVA ===\n")

# Convert to GeneSetCollection
make_gsc <- function(filtered_sets) {
  gsc <- GeneSetCollection()
  for (nm in names(filtered_sets)) {
    gsc <- c(gsc, GeneSet(filtered_sets[[nm]], geneIdType = SymbolIdentifier(), setName = nm))
  }
  return(gsc)
}

run_gsva <- function(gene_expr, filtered_sets, dataset_name) {
  gsc <- make_gsc(filtered_sets)
  cat("Running GSVA for", dataset_name, "...\n")
  
  scores <- gsva(gene_expr, gsc, method = "gsva", kcdf = "Gaussian",
                 rnaseq = FALSE, verbose = FALSE)
  
  cat("GSVA scores dimensions:", dim(scores), "\n")
  return(scores)
}

scores1 <- run_gsva(gene_expr1, filtered_sets1, "GSE124272")
scores3 <- run_gsva(gene_expr3, filtered_sets3, "GSE23130")

# ---- Correlate METRN with signature scores ----
cat("\n=== Correlating METRN with signature scores ===\n")

# GSE124272
cat("\nGSE124272 METRN-Signature correlations:\n")
cor_results1 <- data.frame()
if (!is.null(metrn1)) {
  for (sig_name in rownames(scores1)) {
    cor_val <- cor(metrn1, scores1[sig_name, ], method = "spearman")
    cor_test <- cor.test(metrn1, scores1[sig_name, ], method = "spearman")
    cor_results1 <- rbind(cor_results1, data.frame(
      Dataset = "GSE124272",
      Signature = sig_name,
      Spearman_r = cor_val,
      P_value = cor_test$p.value,
      n_genes_in_set = length(filtered_sets1[[sig_name]])
    ))
    cat(sig_name, ": r =", round(cor_val, 3), "P =", format(cor_test$p.value, digits = 3), "\n")
  }
  
  # FDR correction
  cor_results1$FDR <- p.adjust(cor_results1$P_value, method = "BH")
  cat("\nWith FDR correction:\n")
  print(cor_results1)
}

# GSE23130
cat("\nGSE23130 METRN-Signature correlations:\n")
cor_results3 <- data.frame()
if (!is.null(metrn3)) {
  for (sig_name in rownames(scores3)) {
    cor_val <- cor(metrn3, scores3[sig_name, ], method = "spearman")
    cor_test <- cor.test(metrn3, scores3[sig_name, ], method = "spearman")
    cor_results3 <- rbind(cor_results3, data.frame(
      Dataset = "GSE23130",
      Signature = sig_name,
      Spearman_r = cor_val,
      P_value = cor_test$p.value,
      n_genes_in_set = length(filtered_sets3[[sig_name]])
    ))
    cat(sig_name, ": r =", round(cor_val, 3), "P =", format(cor_test$p.value, digits = 3), "\n")
  }
  
  # FDR correction
  cor_results3$FDR <- p.adjust(cor_results3$P_value, method = "BH")
  cat("\nWith FDR correction:\n")
  print(cor_results3)
}

# ---- Save results ----
cat("\n=== Saving results ===\n")

# Combine correlation results
all_cor <- rbind(cor_results1, cor_results3)
write.csv(all_cor, file.path(out_dir, "signature_correlation_results.csv"), row.names = FALSE)

# Save GSVA scores
write.csv(scores1, file.path(out_dir, "GSVA_scores_GSE124272.csv"))
write.csv(scores3, file.path(out_dir, "GSVA_scores_GSE23130.csv"))

# ---- Visualization ----
cat("\n=== Generating plots ===\n")

# Heatmap of METRN vs signature scores
if (!is.null(metrn1) && nrow(cor_results1) > 0) {
  # Bar plot of correlations
  p <- ggplot(cor_results1, aes(x = reorder(Signature, -Spearman_r), y = Spearman_r, fill = FDR < 0.05)) +
    geom_bar(stat = "identity") +
    geom_hline(yintercept = 0, linetype = "dashed", color = "gray") +
    scale_fill_manual(values = c("TRUE" = "#F44336", "FALSE" = "#9E9E9E"),
                      labels = c("TRUE" = "FDR < 0.05", "FALSE" = "FDR >= 0.05")) +
    labs(title = "METRN-Signature Correlations (GSE124272)",
         x = "Signature", y = "Spearman r",
         fill = "Significance") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  ggsave(file.path(out_dir, "METRN_signature_correlation_GSE124272.png"), p, width = 8, height = 5)
  ggsave(file.path(out_dir, "METRN_signature_correlation_GSE124272.pdf"), p, width = 8, height = 5)
  cat("GSE124272 correlation bar plot saved\n")
}

if (!is.null(metrn3) && nrow(cor_results3) > 0) {
  p <- ggplot(cor_results3, aes(x = reorder(Signature, -Spearman_r), y = Spearman_r, fill = FDR < 0.05)) +
    geom_bar(stat = "identity") +
    geom_hline(yintercept = 0, linetype = "dashed", color = "gray") +
    scale_fill_manual(values = c("TRUE" = "#F44336", "FALSE" = "#9E9E9E"),
                      labels = c("TRUE" = "FDR < 0.05", "FALSE" = "FDR >= 0.05")) +
    labs(title = "METRN-Signature Correlations (GSE23130)",
         x = "Signature", y = "Spearman r",
         fill = "Significance") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  ggsave(file.path(out_dir, "METRN_signature_correlation_GSE23130.png"), p, width = 8, height = 5)
  ggsave(file.path(out_dir, "METRN_signature_correlation_GSE23130.pdf"), p, width = 8, height = 5)
  cat("GSE23130 correlation bar plot saved\n")
}

# ---- METRN-PIEZO1 scatter plot (GSE124272) ----
cat("\n=== Generating METRN-PIEZO1 scatter ===\n")
piezo1_vals <- gene_expr1["PIEZO1", ]
if (!is.null(metrn1) && !is.null(piezo1_vals)) {
  labels1_plot <- rep("Control", ncol(gene_expr1))
  for (i in 1:ncol(gene_expr1)) {
    ti <- pdata1$title[i]
    if (grepl("patient|IDD", ti, ignore.case = TRUE)) labels1_plot[i] <- "IDD"
  }
  
  df_scatter <- data.frame(METRN = metrn1, PIEZO1 = piezo1_vals, Group = labels1_plot)
  
  cor_test <- cor.test(df_scatter$METRN, df_scatter$PIEZO1, method = "spearman")
  
  p <- ggplot(df_scatter, aes(x = METRN, y = PIEZO1, color = Group)) +
    geom_point(size = 4) +
    geom_smooth(method = "lm", se = TRUE, color = "black", linewidth = 0.5) +
    scale_color_manual(values = c("Control" = "#4CAF50", "IDD" = "#F44336")) +
    labs(title = paste0("METRN-PIEZO1 Correlation (GSE124272)\nSpearman r = ", 
                        round(cor_test$estimate, 3), ", P = ", format(cor_test$p.value, digits = 3)),
         x = "METRN Expression", y = "PIEZO1 Expression") +
    theme_bw() + theme(plot.title = element_text(size = 13, face = "bold"))
  
  ggsave(file.path(out_dir, "METRN_PIEZO1_scatter_GSE124272.png"), p, width = 7, height = 5)
  ggsave(file.path(out_dir, "METRN_PIEZO1_scatter_GSE124272.pdf"), p, width = 7, height = 5)
  cat("METRN-PIEZO1 scatter saved\n")
}

cat("\n=== All Step 3 analyses completed ===\n")
cat("Output saved to:", out_dir, "\n")
